Windows build

ericmehl opened the pull request:
This PR gets Gaffer building on Windows MSVC 2017. I've purposefully left out a launching script comparable to the `gaffer` shell script to not give the indication this is a working build, but at least it compiles!

I know this is a pretty big set of commits to review, but I'm hoping many of them are not problematic. But I can break it up if that would be helpful. Submitting them all at once has the advantage of not having to keep multiple branches in sync to verify every change still builds correctly.

### Breaking changes ###

None

### Checklist ###

- [X] I have read the [contribution guidelines](https://github.com/GafferHQ/gaffer/blob/main/CONTRIBUTING.md).
- [X] I have updated the documentation, if applicable.
- [X] I have tested my change(s) in the test suite, and added new test cases where necessary.
- [X] My code follows the Gaffer project's prevailing coding style and conventions.

ericmehl commented:
A particularly sticky one is probably https://github.com/GafferHQ/gaffer/pull/4570/commits/cd2b706af6de5f10c8f9793c5bfe1cf4ed04c580, from John's comment in the preceding version. I'm not sure how to make it much faster and happy for ideas. 

With some text filtering of each line, could it possibly be multithreaded? The steps of creating all nodes, setting values and connecting nodes seem like they could be made parallel?

johnhaddon commented:
> I know this is a pretty big set of commits to review, but I'm hoping many of them are not problematic. But I can break it up if that would be helpful.

As a compromise, would it be OK if I cherry-picked all the no-brainers over to `main`, and then we can rebase the PR and focus in on what remains? I think you're right that we're going to need to spend a little bit of time on the script loading code...

ericmehl commented:
That sounds perfect, cherry-pick away.

ericmehl commented:
Rebasing pulled in the Arnold `filterMetadata` build step in Scons, which I'm pretty sure is causing the new failure when it uses a modified `PATH` in the supplied environment. That doesn't work on Windows because it would just be too easy. I'll check that that's the problem and sort it out first thing tomorrow.

ericmehl commented:
`filterMetadata` is working now on Windows, and it looks like the Mac failure is the same one fixed in #4575. Ready for review.

johnhaddon commented on src/GafferUI/ViewportGadget.cpp:
```diff
@@ -1911,7 +1911,7 @@ void ViewportGadget::SelectionScope::begin( const ViewportGadget *viewportGadget
 	/// \todo It would be better to base this on whether we have a depth buffer or not, but
 	/// we don't have access to that information right now.
-	m_depthSort = camera->isInstanceOf( IECoreGL::PerspectiveCamera::staticTypeId() );
+	m_depthSort = camera->isInstanceOf( IECoreGL::Camera::staticTypeId() );
```
This is interesting @danieldresser-ie. In https://github.com/ImageEngine/cortex/commit/cecdf79a5a61f5821a099e355de6f9a4b1cb8ea2 we removed `IECoreGL::PerspectiveCamera`, but forgot to remove the header. GCC and Clang let us get away with it but MSVC has pointed it out.

So I think `m_depthSort` has been unconditionally `false` ever since. I don't recall the selection code that well, but i think perhaps this would only have caused a problem with the legacy `GL_SELECT` mechanism, because the ID-based one will only see the frontmost object anyway. And I think this check was really a hack to determine if we were in the Viewer (where we have a depth buffer and want depth sorting) or the GraphEditor (where we don't).

If that's correct @ericmehl, then this commit will make `m_depthSort` unconditionally `true` instead. We'll need to do some investigation to see if that's problematic or not...

johnhaddon commented on src/GafferTest/ContextTest.cpp:
```diff
@@ -238,7 +246,11 @@ std::tuple<int,int,int,int> GafferTest::countContextHash32Collisions( int contex
 			for( int j = 0; j < 20; j++ )
 			{
-				c.set( numberNames[j], rand_r( &rand_seed ) );
+#ifndef _MSC_VER
+					c.set( numberNames[j], rand_r( &rand_seed ) );
+#else
+					c.set( numberNames[j], rand() );
+#endif
```
Can we just use a more recent random generator that works cross-platform? One from Imath or one of the new `std` ones? I'd like to limit the `#ifdefing` to the minimum necessary.

ericmehl replied:
fixed in 6beb287da180b823dac5a828d419c55cdc9fd282

johnhaddon commented on src/GafferModule/ScriptNodeBinding.cpp:
```diff
@@ -119,78 +86,27 @@ const std::string formattedErrorContext( int lineNumber, const std::string &cont
-
-	// Parse the whole script, getting an abstract syntax tree for a
-	// module which would execute everything.
-	mod_ty mod = PyParser_ASTFromString(
```
My recollection is that this `PyParser_ASTFromString()` approach was quicker than the manual line-by-line splitting approach, and script load speed is a pretty important metric for us. I think the best we can do for now is `#ifdef` for Windows so that performance doesn't suffer on Linux.

It looks like this has all changed in Python 3.9, so I'll be forced to revisit it soon enough anyway. At that point I'll do some benchmarking and try to come up with a single performant cross-platform approach.

ericmehl replied:
I added the AST code back in 4dfab183abf98c32f1643295aa58f0dfcd468e3c

johnhaddon commented on src/GafferTestModule/SignalsTest.cpp:
```diff
@@ -94,12 +94,20 @@ void testDisconnectMatchingLambda()
 	auto connection2 = signal.connect( slot2 );
 	GAFFERTEST_ASSERTEQUAL( signal.numSlots(), 2 );
 
+#ifdef _MSC_VER
+	signal.disconnect( static_cast<void( __cdecl * )()>( slot1 ) );
+#else
```
Ugh! I'm way out of my depth here, but it would be great if we could fix this on the Signals side, rather than require client code to do the weird cast. I'm pretty sure this test was OK with `boost::signals`, so presumably there's some voodoo we could put inside `disconnect()` to make this possible?

johnhaddon replied:
Scratch that - this one is definitely my problem and not yours. Could you just add a todo saying "Can this be dealt with internally in `Signal::disconnect()?"?

ericmehl replied:
Of course. If it's not too much of a distraction, what led you to that conclusion? This one has made it into my top five "why can't Windows just cooperate" list and I don't 100% understand why it can't do the comparison itself.

johnhaddon replied:
> what led you to that conclusion?

It's certainly not based on any belief that I know enough to fix it! I just figured that the ideal fix is inside `Signal` and that your mission to get Gaffer on Windows didn't really need to get held up by it...

ericmehl replied:
I added the todo in 754fe935bdee94c056df89cefadbbfc7044ab130

johnhaddon reviewed (changes requested):
Thanks Eric! I've added my thoughts inline...

ericmehl commented on src/GafferModule/GafferModule.cpp:
```diff
@@ -165,12 +167,15 @@ void clobberArgv()
 {
 	// Some things (for instance, `ps` in default mode) look at `argv` to get
 	// the name.
+#ifndef _MSC_VER
 	clobberArgv();
+#endif
```
I haven't been able to dig anything up yet, but I want to keep looking. I tried [SetProcTitle](https://github.com/dvarrazzo/py-setproctitle) which is linked from a Python mailing list discussion about this topic, but it doesn't do anything on Windows.

ericmehl commented on .github/workflows/main/installDependencies.py:
```diff
@@ -87,7 +87,11 @@
+	subprocess.check_call(
+		[ "7z", "x", archiveFileName, "-o{}".format( args.dependenciesDir ), "-aoa", "-y" ],
+		stdout = subprocess.DEVNULL,
+		stderr = subprocess.DEVNULL
```
Fixed in cdace24ccb95743d5731e88178002091b02a83d9

ericmehl commented on .github/workflows/main.yml:
```diff
@@ -141,6 +134,10 @@ jobs:
         Xvfb :99 -screen 0 1280x1024x24 &
         metacity&
         useradd -m testUser
+        echo $PATH > $GITHUB_PATH
```
Fixed in 16308057b9fc6938236475a09018be2dac25feaa

ericmehl commented on .github/workflows/main.yml:
```diff
@@ -198,21 +195,31 @@ jobs:
+          arnoldRoot = os.path.join( os.environ["GITHUB_WORKSPACE"], "arnoldRoot", arnoldVersion )
+          os.environ["ARNOLD_ROOT"] =  arnoldRoot
+
+          subprocess.check_call( sys.executable + " .github/workflows/main/installArnold.py --version " + arnoldVersion, shell = True )
```
Fixed in 07ed5448b513d801436ce90ab3a95da63dcc5c01

ericmehl commented on .github/workflows/main/setBuildVars.py:
```diff
@@ -171,3 +171,7 @@
 	print( "Setting $GAFFER_GITHUB_RELEASEID to '%s'" % releaseId )
 	f.write( 'GAFFER_GITHUB_RELEASEID=%s\n' % releaseId )
+
+	if os.name != "nt":
```
Fixed in 89613a14ac75fb33cb312dad4d6a73d9283b9c15

ericmehl commented on include/GafferBindings/SignalBinding.h:
```diff
@@ -38,13 +38,22 @@
 
+#include "Gaffer/Signals.h"
+
+#include "boost/python.hpp"
```
Fixed in db2c9117d9165ddeb0fb0e39923265362e2c0ffb

ericmehl commented on include/GafferBindings/SignalBinding.h:
```diff
@@ -38,13 +38,22 @@
+
+	GAFFERBINDINGS_API boost::python::object pythonConnection( const Gaffer::Signals::Connection &connection, const boost::python::object &scoped );
+
+} // namespace Detail
```
I think my brain came apart on that one, it's not necessary to change it. I backed that change out in 0b64bc3dabd33fcf600a0bc0f213bfd0193648ea

ericmehl commented on include/GafferArnoldUI/ArnoldLightVisualiser.h:
```diff
@@ -0,0 +1,74 @@
+namespace GafferArnoldUI
+{
+
+class GAFFERARNOLDUI_API ArnoldLightVisualiser : public GafferSceneUI::StandardLightVisualiser
```
> I'm wondering if the stray GAFFERSCENEUI_API in ArnoldLightVisualiser.cpp is what sent you down this path?
Yeah, I think that's what sent me down that path. I reverted it in 6422134eddb90dc9a692f3d88c2d7b1a92ba89ea and removed the unnecessary export in 8498bd8aace7e5237bb80f2da1e858063d1e0c65

ericmehl commented on include/GafferDelight/IECoreDelightPreview/NodeAlgo.h:
```diff
@@ -55,10 +57,10 @@ namespace NodeAlgo
 /// NSI node with the specified handle, returning true on
 /// success and false on failure.
-bool convert( const IECore::Object *object, NSIContext_t context, const char *handle );
+GAFFERDELIGHT_API bool convert( const IECore::Object *object, NSIContext_t context, const char *handle );
```
Nope, you didn't miss anything. I guess once I get to exporting symbols, it's hard to stop. Reverted in ab39f98d6170f1fc44dd9a7422928eee35874194

ericmehl commented on src/Gaffer/ScriptNode.cpp:
```diff
@@ -970,9 +968,18 @@ void ScriptNode::plugSet( Plug *plug )
+			std::ofstream testOpen( fileName.c_str(), std::fstream::app );
+			isReadOnly = !testOpen.is_open();
+			testOpen.close();
+		}
```
I didn't know about `_access`, that's a much better solution. I went with the `#define` method since I also am declaring the `W_OK`. Happy to use `inline` if that's better. Fixed in 9367ec0436ff84538ccfa1262ad157eb7e726dd6

ericmehl commented on include/Gaffer/Process.h:
```diff
@@ -39,6 +39,8 @@
 #include "Gaffer/Export.h"
 #include "Gaffer/ThreadState.h"
+#include "Gaffer/Plug.h"
+#include "Gaffer/Context.h"
```
Fixed in 1c13279459dcba2b8f6e5070135843d564571786

ericmehl commented on src/IECoreArnold/Renderer.cpp:
```diff
@@ -960,7 +960,7 @@ class ArnoldAttributes : public IECoreScenePreview::Renderer::AttributesInterfac
 				}
 
-				if( it->first.string() == g_arnoldLightFilterShaderAttributeName )
+				if( it->first.string() == g_arnoldLightFilterShaderAttributeName.string() )
```
Fixed in 9f90b5d25a2c961849debd4eaea9b51d262606ad

ericmehl commented:
I added new commits and reverted some old ones to address the notes. I can squash and rebase as well if it makes it easier to review.

I addressed all of the notes except for https://github.com/GafferHQ/gaffer/pull/4570#discussion_r808312444 about the depth buffer test. I'm happy to investigate further, but I'm not sure what the desired effect of the `sort` vs. `reverse` switch is meant to be at https://github.com/GafferHQ/gaffer/blob/40b2731b311155dbe99f739f2ea460c7296ff505/src/GafferUI/ViewportGadget.cpp#L1935-L1942
where it's been always using the `reverse` branch since `PerspectiveCamera` was removed.

johnhaddon commented:
> I added new commits and reverted some old ones to address the notes. I can squash and rebase as well if it makes it easier to review.

Thanks! I've reviewed them as they are, but it'd be great if you could squash the fixes now in preparation to merge.

> I'm happy to investigate further, but I'm not sure what the desired effect of the sort vs. reverse switch is meant to be

I _think_ it's something like this, but my memory is a bit hazy...`IECoreGL::Selector` has three modes : `GLSelect`, `OcclusionQuery` and `IDRender`. `GL_Select` is the original one that returns all the objects in the selection region, in the order they are drawn. Therefore, if we are using a depth buffer, we need to sort by depth to choose the frontmost one. If we're not using a depth buffer, we just want to take the last one. But in practice the underlying `GL_SELECT` mechanism it uses is deprecated and crazy slow on modern NVIDIA drivers, so we avoid it in favour of the second two. For the second two, I don't think it matters what we do - `OcclusionQuery` doesn't provide depth in any case, and `IDRender` looks in an ID buffer so can only see the topmost object anyway.

So, I think the question is "do we use `GL_SELECT` anywhere any more?". It's the default for `ViewportGadget::SelectionScope` but it looks like all the clients of that class choose a different type explicitly (we should change the default). But I _think_ there's also a fallback mode to GL_SELECT that might be active on Mac - I've got a Mac build going now to try to figure that out.

johnhaddon commented:
> But I think there's also a fallback mode to GL_SELECT that might be active on Mac - I've got a Mac build going now to try to figure that out.

So, yep, we do fall back to GL_SELECT on Mac. And we're not sorting in `ViewportGadget::SelectionScope` because the cast to PerspectiveCamera will always fail. But selection in the Viewer still works, because `SceneGadget::objectAt()` is doing its own sorting. What a mess! Not sure what's best in terms of the Windows PR. Maybe hardcode `m_depthSort = false` and add a todo?

ericmehl commented:
> Maybe hardcode m_depthSort = false and add a todo?

That sounds good to me, especially since this PR is already substantial. I noted it in https://github.com/GafferHQ/gaffer/pull/4570/commits/9fd66d39c33dfc3f6a34923f456fad02b88157f3

ericmehl commented:
I squashed a bunch of commits and have a more manageable history, ready for review.

johnhaddon commented:
Closing now that #4594 is merged. Thanks for your patience on this Eric...
